Fixes #37347 - Add Packages wizard to All Hosts page

jeremylenz opened the pull request:
Requires https://github.com/theforeman/foreman/pull/10128

#### What are the changes introduced in this pull request?

Add the Packages Wizard to the new All Hosts page, based on designs.

![image](https://github.com/Katello/katello/assets/22042343/f06cfda7-824a-49d9-b544-1aeab5f2b2c8)

There are 4 steps in the wizard:

1. Select action
2. Select packages to update / install
3. Review hosts
4. Review

* Based on your selected action, step 2 may or may not be shown

And there are 3 flows:

1. Update all packages
2. Update packages
3. Install packages


This also required [adding features in Foreman React components](https://github.com/theforeman/foreman/pull/10128), and adding new capabilities here:

- Add `:distinct` and `:select` options to `api_controller.rb` so that you can get back a list distinct by package name. I had to write some SQL fragments myself since Active Record wasn't playing nice with `.count`.
- Add errors in `package_names_for_job_template` if the package search turns up empty. This is to prevent the template from updating all packages when you only wanted to update one package.
- Extend RemoteExecutionActions to handle host searches instead of just single hostnames
- Extend customizedRexUrlHelpers to handle host searches instead of just single hostnames

#### Considerations taken when implementing this change?

For the wizard itself, I used the ["React next" implementation](https://v4-archive.patternfly.org/v4/components/wizard/react-next) in PF4. I thought this was a much cleaner API because of the handy wizardContext it provides, and also thought it might be easier to migrate once we move to PF5.

For most of the wizard, I use a React Context, `BulkPackagesWizardContext`, to share values between all the wizard steps and components. But the `<HostReview>` component should be reusable in other wizards, and using context would have made it nonreusable. So for `<HostReview>` I made sure not to use any context values, and instead passed all values through props.

Originally I had the `useBulkSelect` calls within each separate component (`BulkPackagesTable` and `HostReview`). And I was trying to share values by adding state to the top-level context. But I quickly realized that was a bad decision. Maybe surprisingly, things got a lot cleaner and more elegant once I moved both `useBulkSelect`s to the top level.

#### What are the testing steps for this pull request?

1. Test all the flows!!
2. Make sure all the wizard step validations make sense
3. When you're on step 4 (Review), make sure the "Edit" buttons both take you back to the appropriate step, and remember your selections
4. Test both "via remote execution" and "via customized remote execution". The "via remote execution" should poll the job and give you a toast notification when the job is complete. "Via customized remote execution" will take you to the job wizard, where you can verify the host and package inputs are correct.
5. Test other pages that use TableIndexPage, like HW Models, and make sure they still work

ianballou reviewed (commented):
Seeing this following when selecting step 2 of the wizard without editing anything in step 1:
![image](https://github.com/Katello/katello/assets/14796566/98922972-de19-4378-ad2b-9553b2ff08a1)

```
TypeError: isSelected is not a function
    in RowSelectTd (created by Table)
    in tr (created by TrBase)
    in TrBase (created by Tr)
    in Tr (created by Table)
    in tbody (created by TbodyBase)
    in TbodyBase (created by Tbody)
    in Tbody (created by Table)
    in table (created by TableComposableBase)
    in TableComposableBase (created by TableComposable)
    in TableComposable (created by Table)
    in Table (created by TableIndexPage)
    in section (created by PageSection)
    in PageSection (created by TableIndexPage)
    in div (created by TableIndexPage)
    in TableIndexPage (created by HostReview)
    in HostReview (created by BulkPackagesWizard)
    in div (created by WizardBody)
    in div (created by WizardBody)
    in WizardBody (created by WizardToggle)
    in div (created by WizardToggle)
    in div (created by WizardToggle)
    in WizardToggle (created by WizardInternal)
    in WizardInternal (created by Wizard)
    in div (created by Wizard)
    in WizardContextProvider (created by Wizard)
    in Wizard (created by BulkPackagesWizard)
    in BulkPackagesWizard (created by BulkPackagesWizardModal)
    in div (created by ModalBox)
    in ModalBox (created by ModalContent)
    in div (created by FocusTrap)
    in FocusTrap (created by ForwardRef)
    in ForwardRef (created by ModalContent)
    in div (created by Backdrop)
    in Backdrop (created by ModalContent)
    in ModalContent (created by Modal)
    in Modal (created by BulkPackagesWizardModal)
    in BulkPackagesWizardModal
    in Slot (created by ConnectFunction)
    in ConnectFunction (created by HostsIndex)
    in section (created by PageSection)
    in PageSection (created by TableIndexPage)
    in div (created by TableIndexPage)
    in TableIndexPage (created by HostsIndex)
    in HostsIndex (created by Context.Consumer)
    in Route (created by AppSwitcher)
    in Switch (created by ForemanSwitcher)
    in ForemanSwitcher (created by AppSwitcher)
    in AppSwitcher (created by ReactApp)
    in ErrorBoundary (created by ReactApp)
    in main (created by Page)
    in div (created by Page)
    in Page (created by Layout)
    in div (created by FlexItem)
    in FlexItem (created by Layout)
    in div (created by Flex)
    in Flex (created by Layout)
    in Layout (created by ConnectedLayout)
    in ConnectedLayout (created by ReactApp)
    in Router (created by ConnectedRouter)
    in ConnectedRouter (created by Context.Consumer)
    in ConnectedRouterWithContext (created by ConnectFunction)
    in ConnectFunction (created by ReactApp)
    in ApolloProvider (created by ReactApp)
    in div (created by ReactApp)
    in ReactApp (created by I18nProviderWrapper(ReactApp))
    in IntlProvider (created by I18nProviderWrapper(ReactApp))
    in I18nProviderWrapper(ReactApp) (created by StoreProvider(I18nProviderWrapper(ReactApp)))
    in Provider (created by StoreProvider(I18nProviderWrapper(ReactApp)))
    in StoreProvider(I18nProviderWrapper(ReactApp)) (created by DataProvider(StoreProvider(I18nProviderWrapper(ReactApp))))
    in DataProvider(StoreProvider(I18nProviderWrapper(ReactApp)))
    in AwaitedMount
```

ianballou reviewed (commented):
The "Cancel" button seems to be non-functional

jeremylenz commented:
> The "Cancel" button seems to be non-functional

oh yeah I've been meaning to fix that :smile: I think it works on the review page at least

jeremylenz commented:
> Seeing this following when hitting Next for all options except "Upgrade packages":

Do you have the Foreman PR checked out?

ianballou commented:
> Do you have the Foreman PR checked out?

That must be it :)

ianballou reviewed (commented):
Packages here seem to be referred to by name alone. Users would probably like to select and see the version that they're installing or upgrading to. 
Could packages be referred to by their full NEVRA? Or, perhaps the package selector could mirror what's on the host details page:

![image](https://github.com/Katello/katello/assets/14796566/afc21548-b57f-45b7-bd6f-c8fd718beb2b)

ianballou commented:
Looks like users can currently filter package versions by using the search bar, but I'm guessing they'd like to be able to see what versions are available.

jeremylenz commented:
> Looks like users can currently filter package versions by using the search bar, but I'm guessing they'd like to be able to see what versions are available.

This bulk wizard only allows upgrading to the latest version. I can't think of a sane way to allow any granularity there, when you're dealing with multiple hosts that may have different versions installed and different versions available.

jeremylenz commented:
> Users would probably like to select and see the version that they're installing or upgrading to.

That would be nice! But the version they're installing or upgrading to could vary from host to host, depending on that host's LCE/CV.

ianballou commented:
Hit an error trying to install 4 RPMs on two hosts:

```
2024-05-02T17:26:03 [I|app|1d601299] Backtrace for 'Action failed' error (ActiveRecord::RecordNotSaved): Validation failed: Targeting: Bookmark Must select a bookmark or enter a search query, Search query Must select a bookmark or enter a search query
 1d601299 | /home/vagrant/foreman_remote_execution/app/models/job_invocation_composer.rb:460:in `save!'
 1d601299 | /home/vagrant/foreman_remote_execution/app/models/job_invocation_composer.rb:433:in `trigger'
 1d601299 | /home/vagrant/foreman_remote_execution/app/models/job_invocation_composer.rb:441:in `trigger!'
 1d601299 | /home/vagrant/foreman_remote_execution/app/controllers/api/v2/job_invocations_controller.rb:93:in `create'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/abstract_controller/base.rb:228:in `process_action'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_controller/metal/rendering.rb:30:in `process_action'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/abstract_controller/callbacks.rb:42:in `block in process_action'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/callbacks.rb:117:in `block in run_callbacks'
 1d601299 | /home/vagrant/foreman/app/controllers/concerns/foreman/controller/timezone.rb:10:in `set_timezone'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/callbacks.rb:126:in `block in run_callbacks'
 1d601299 | /home/vagrant/foreman/app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/callbacks.rb:126:in `block in run_callbacks'
 1d601299 | /home/vagrant/foreman/app/controllers/concerns/foreman/controller/topbar_sweeper.rb:12:in `set_topbar_sweeper_controller'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/callbacks.rb:126:in `block in run_callbacks'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/audited-5.6.0/lib/audited/sweeper.rb:16:in `around'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/callbacks.rb:126:in `block in run_callbacks'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/audited-5.6.0/lib/audited/sweeper.rb:16:in `around'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/callbacks.rb:126:in `block in run_callbacks'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/callbacks.rb:137:in `run_callbacks'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/abstract_controller/callbacks.rb:41:in `process_action'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_controller/metal/rescue.rb:22:in `process_action'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_controller/metal/instrumentation.rb:34:in `block in process_action'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/notifications.rb:203:in `block in instrument'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/notifications.rb:203:in `instrument'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_controller/metal/instrumentation.rb:33:in `process_action'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_controller/metal/params_wrapper.rb:249:in `process_action'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activerecord-6.1.7.7/lib/active_record/railties/controller_runtime.rb:27:in `process_action'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/abstract_controller/base.rb:165:in `process'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionview-6.1.7.7/lib/action_view/rendering.rb:39:in `process'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_controller/metal.rb:190:in `dispatch'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_controller/metal.rb:254:in `dispatch'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/routing/route_set.rb:50:in `dispatch'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/routing/route_set.rb:33:in `serve'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/routing/mapper.rb:19:in `block in <class:Constraints>'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/routing/mapper.rb:49:in `serve'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/journey/router.rb:50:in `block in serve'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/journey/router.rb:32:in `each'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/journey/router.rb:32:in `serve'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/routing/route_set.rb:842:in `call'
 1d601299 | /home/vagrant/katello/lib/katello/middleware/organization_created_enforcer.rb:18:in `call'
 1d601299 | /home/vagrant/katello/lib/katello/middleware/event_daemon.rb:10:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/static.rb:24:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/static.rb:24:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/apipie-dsl-2.6.2/lib/apipie_dsl/static_dispatcher.rb:67:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/apipie-rails-1.3.0/lib/apipie/static_dispatcher.rb:74:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/bullet-7.1.6/lib/bullet/rack.rb:17:in `call'
 1d601299 | /home/vagrant/foreman/lib/foreman/middleware/libvirt_connection_cleaner.rb:9:in `call'
 1d601299 | /home/vagrant/foreman/lib/foreman/middleware/telemetry.rb:10:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/apipie-rails-1.3.0/lib/apipie/middleware/checksum_in_headers.rb:27:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-openid-1.4.2/lib/rack/openid.rb:98:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/tempfile_reaper.rb:15:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/etag.rb:27:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/conditional_get.rb:40:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/head.rb:12:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/http/permissions_policy.rb:22:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/http/content_security_policy.rb:19:in `call'
 1d601299 | /home/vagrant/foreman/lib/foreman/middleware/logging_context_session.rb:22:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/session/abstract/id.rb:266:in `context'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/session/abstract/id.rb:260:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/cookies.rb:697:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activerecord-6.1.7.7/lib/active_record/migration.rb:601:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/callbacks.rb:98:in `run_callbacks'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/callbacks.rb:26:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/executor.rb:14:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/actionable_exceptions.rb:18:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/debug_exceptions.rb:29:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/railties-6.1.7.7/lib/rails/rack/logger.rb:37:in `call_app'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/railties-6.1.7.7/lib/rails/rack/logger.rb:28:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/sprockets-rails-3.4.2/lib/sprockets/rails/quiet_assets.rb:13:in `call'
 1d601299 | /home/vagrant/foreman/lib/foreman/middleware/logging_context_request.rb:11:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/request_id.rb:26:in `call'
 1d601299 | /home/vagrant/katello/lib/katello/prevent_json_parsing.rb:12:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/method_override.rb:24:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/runtime.rb:22:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/activesupport-6.1.7.7/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/executor.rb:14:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/static.rb:24:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/sendfile.rb:110:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/actionpack-6.1.7.7/lib/action_dispatch/middleware/host_authorization.rb:148:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/secure_headers-6.5.0/lib/secure_headers/middleware.rb:11:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/railties-6.1.7.7/lib/rails/engine.rb:539:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/railties-6.1.7.7/lib/rails/railtie.rb:207:in `public_send'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/railties-6.1.7.7/lib/rails/railtie.rb:207:in `method_missing'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/urlmap.rb:74:in `block in call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/urlmap.rb:58:in `each'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/rack-2.2.9/lib/rack/urlmap.rb:58:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/puma-6.4.2/lib/puma/configuration.rb:272:in `call'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/puma-6.4.2/lib/puma/request.rb:100:in `block in handle_request'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/puma-6.4.2/lib/puma/thread_pool.rb:378:in `with_force_shutdown'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/puma-6.4.2/lib/puma/request.rb:99:in `handle_request'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/puma-6.4.2/lib/puma/server.rb:464:in `process_client'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/puma-6.4.2/lib/puma/server.rb:245:in `block in run'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/puma-6.4.2/lib/puma/thread_pool.rb:155:in `block in spawn_thread'
 1d601299 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/logging-2.3.1/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
2024-05-02T17:26:03 [D|app|1d601299]   Rendering layout api/v2/layouts/error_layout.json.erb
2024-05-02T17:26:03 [D|app|1d601299]   Rendering api/v2/errors/standard_error.json.rabl within api/v2/layouts/error_layout
2024-05-02T17:26:03 [I|app|1d601299]   Rendered api/v2/errors/standard_error.json.rabl within api/v2/layouts/error_layout (Duration: 1.3ms | Allocations: 300)
2024-05-02T17:26:03 [I|app|1d601299]   Rendered layout api/v2/layouts/error_layout.json.erb (Duration: 1.8ms | Allocations: 488)
2024-05-02T17:26:03 [I|app|1d601299] Completed 500 Internal Server Error in 105ms (Views: 5.9ms | ActiveRecord: 24.3ms | Allocations: 43711)

```

ianballou reviewed (commented):
Confirmed that updating all & selected packages works fine. Also confirmed that the edit buttons work, along with run via rex and customized rex.

ianballou commented:
> `Must select a bookmark or enter a search query`
> 
> @ianballou Can you choose 'via customized' and see what you end up with for the host search query? Does it resolve to 0 or 2 hosts?

![image](https://github.com/Katello/katello/assets/14796566/1c248a0e-c922-4aa7-b346-2cb017045fdc)

ianballou commented:
The bookmark issue does not appear for custom REX.

jeremylenz commented:
@ianballou Bookmark error issue should be fixed now.

ianballou commented:
Bookmark error is fixed, up next:

```
Failed to initialize: RuntimeError - Failed rendering template: error during rendering: No available packages found for search term 'name ^ (vim-minimal,vim-filesystem,vim-enhanced,vim-common)'. Check the host's content view environments.
```

Confirmed vim-enhanced is available.

ianballou commented:
I just tested:
1) Update yum on one host but not another
2) Update yum via the UI
3) On the host with latest yum, all packages got updated

ianballou commented:
Re: showing full NEVRAs, it sounds like that wasn't in the original design, so I'm cool with keeping that as a potential future RFE. This page is mainly for quickly firing off commands to many hosts, so if they need full NEVRAs they can make a custom REX job easily anyway.

jeremylenz commented:
@ianballou ok, try the already-upgraded package again. Added a new check that should catch it.

ianballou commented:
Noticed one thing -- when you do an action, the kebab menu is disabled until the page is reloaded

ianballou commented:
If you select 'via customized remote execution', the upgrade button grows:

![image](https://github.com/Katello/katello/assets/14796566/2ca3218e-7786-4018-92cd-fb858c0e4a6c)

vs

![image](https://github.com/Katello/katello/assets/14796566/d5039451-a24c-4ec1-b407-3117f2e7ff66)

ianballou commented:
I also noticed that if you select some packages for upgrade, then go back and change the action from 'update some' to 'update all', the selected packages carry into the job template at the end.

ianballou commented:
Besides the comments above, each action seems to be functional now.

jeremylenz commented:
> Noticed one thing -- when you do an action, the kebab menu is disabled until the page is reloaded

I'm not seeing this. Any more details? Which kebab?

> If you select 'via customized remote execution', the upgrade button grows:

Fixed :+1: 

> I also noticed that if you select some packages for upgrade, then go back and change the action from 'update some' to 'update all', the selected packages carry into the job template at the end.

Nice catch!! Fixed :+1: 

Also, the Cancel buttons should work now.

ianballou commented:
I can't reproduce the disabled kebab issue now :+1:

ianballou commented:
![image](https://github.com/Katello/katello/assets/14796566/55d2110c-41c2-4176-a185-dc76297ac97c)

The wasn't supposed to be a dropdown to review the targeted hosts, was there?

ianballou commented:
When updating packages on a host that doesn't need the packages updated, I see:

```
Warning! error during rendering: unexpected token at '$USER_INPUT[Selected update versions]'
^[[32m2024-05-03T15:36:08^[[0m [^[[32mI^[[0m|^[[36mapp^[[0m|62236b82] Backtrace for 'Error during rendering of a job template' error (JSON::ParserError): unexpected token at '$USER_INPUT[Selected update versions]'
 62236b82 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/json-2.7.2/lib/json/common.rb:220:in `parse'
 62236b82 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/json-2.7.2/lib/json/common.rb:220:in `parse'
 62236b82 | /home/vagrant/katello/app/models/katello/concerns/host_managed_extensions.rb:525:in `yum_names_for_job_template'
 62236b82 | /home/vagrant/katello/app/models/katello/concerns/host_managed_extensions.rb:502:in `package_names_for_job_template'

```

ianballou commented:
Also, wanted to check, when updating packages, `dnf` is receiving specific versions rather than names:

![image](https://github.com/Katello/katello/assets/14796566/15b62311-0f44-4172-8afe-6485b979ccbf)

Is that intentional?

jeremylenz commented:
> There wasn't supposed to be a dropdown to review the targeted hosts like with packages, was there?

Nope, I made it according to the mockup. I guess you can click Edit if you want to see your hosts again :shrug:

jeremylenz commented:
> Also, wanted to check, when updating packages, `dnf` is receiving specific versions rather than names:

Actually after thinking about it I think this is okay. This is because by the time we get to `yum_names_for_job_template`, we're in the context of a single host. Our `search` is something like `name ^ (vim,vim-minimal)`, and we do

```rb
pkg_name_archs = installed_packages.search_for(search).distinct.pluck(:name, :arch)
```
```rb
upgrades = ::Katello::Rpm.installable_for_hosts([self]). ...
```

And then we compare `upgrades` to `pkg_name_archs`, and that's where we get those version numbers. In other words, the package versions you see in the rendered template should be the latest installable version _for that host_.

jeremylenz commented:
> When updating packages on a host that doesn't need the packages updated, I see:
> 
> ```
> Warning! error during rendering: unexpected token at '$USER_INPUT[Selected update versions]'
> ^[[32m2024-05-03T15:36:08^[[0m [^[[32mI^[[0m|^[[36mapp^[[0m|62236b82] Backtrace for 'Error during rendering of a job template' error (JSON::ParserError): unexpected token at '$USER_INPUT[Selected update versions]'
>  62236b82 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/json-2.7.2/lib/json/common.rb:220:in `parse'
>  62236b82 | /home/vagrant/foreman/.vendor/ruby/3.0.0/gems/json-2.7.2/lib/json/common.rb:220:in `parse'
>  62236b82 | /home/vagrant/katello/app/models/katello/concerns/host_managed_extensions.rb:525:in `yum_names_for_job_template'
>  62236b82 | /home/vagrant/katello/app/models/katello/concerns/host_managed_extensions.rb:502:in `package_names_for_job_template'
> ```

It's so weird.. The "versions" input is marked as not required, but it still acts as if it's required. I guess I could have changed the locked job template to check for its presence, but I found it easier to add `?? []` in the JavaScript.

Now you should instead get

```
Warning! error during rendering: No upgradable packages found for search term 'name ^ (curl)'. The host may already have the latest version(s) installed.
```

ianballou commented on webpack/components/extensions/HostDetails/Tabs/RemoteExecutionActions.js:
```diff
@@ -67,16 +74,22 @@ const katelloPackageUpdateParams = ({ hostname, packageName, descriptionFormat }
+    hostname, search, hostSearch, versions, descriptionFormat,
+  } = options;
+  const searchQuery = hostSearch ?? `name ^ (${hostname})`;
+  console.log(versions)
```
console.log

ianballou commented:
Things all seem to be working now!

m-bucher commented:
![image](https://github.com/Katello/katello/assets/1557314/bc4d8cdc-d844-444c-b545-ad1c2348596c)
Bookmark-menu within wizard is clipped to wizard window.

I must admit it took me some time to find the wizard :see_no_evil:

ianballou commented:
Looks like the bookmark clipping issue only happens when your screen is pretty narrow at least. But having it work with the side scroller would be cool if possible.

ianballou commented:
> Also, `Update Packages` package selection only shows RPMs (/packages) and not DEBs (/debs) 😁

I think we may need some help here on this part like with host details page :)

jeremylenz commented:
> Bookmark-menu within wizard is clipped to wizard window.

Flipped the bookmark dropdown so it's entirely within the wizard window. This should fix the issue.  (Will require updating both Foreman and Katello PR branches.)

evgeni commented:
/packit build

jeremylenz commented:
/packit build

evgeni commented:
RPMs were built successfully.  \o/

ianballou commented:
Still working just fine for me!

parthaa commented on app/controllers/katello/api/v2/api_controller.rb:
```diff
@@ -88,6 +88,13 @@ def scoped_search(query, default_sort_by, default_sort_order, options = {})
       query = query.order("#{query.table_name}.id DESC") unless group #secondary order to ensure sort is deterministic
       query = query.includes(includes) if includes.length > 0
 
+      if params[:distinct] && options[:select]
```
can you add an api controllers test to test this.

parthaa commented on app/controllers/katello/api/v2/packages_controller.rb:
```diff
@@ -43,7 +43,17 @@ def auto_complete_arch
     param_group :search, ::Katello::Api::V2::ApiController
     def index
-      super
+      if params[:distinct]
```
can you add an api controllers test to test this.

parthaa commented on app/models/katello/concerns/host_managed_extensions.rb:
```diff
@@ -503,11 +503,16 @@ def package_names_for_job_template(action:, search:, versions: nil)
-          ::Katello::Rpm.yum_installable_for_host(self).search_for(search).distinct.pluck(:name)
+          yum_installable = ::Katello::Rpm.yum_installable_for_host(self).search_for(search).distinct.pluck(:name)
+          if yum_installable.empty?
+            fail _("No available packages found for search term '%s'. Check the host's content view environments and already-installed packages.") % search
```
```suggestion
            fail n_("No available packages found for search term '%s'. Check the host's content view environments and already-installed packages.") % search
```

parthaa reviewed (approved):
APJ. Nice work :100:

jeremylenz commented on .github/workflows/ruby.yml:
```diff
@@ -25,3 +25,4 @@ jobs:
       plugin: katello
       postgresql_container: ghcr.io/theforeman/postgresql-evr
       test_existing_database: false
+      foreman_version: refs/pull/10128/head
```
Added this to run tests with the Foreman PR. Tests are green, except that our "React Tests" check seems to ignore the Foreman version specified.  However, `npm run test` passes locally (with the Foreman PR checked out):

```
Test Suites: 137 passed, 137 total
Tests:       584 passed, 584 total
Snapshots:   165 passed, 165 total
```

MariaAga commented:
If there are no packages, step 2 should be invalid (probably?) and step 3 should not be available

jeremylenz commented:
@MariaAga Updated to invalidate the package select step when there are no upgradable packages in the system. Also added an alert.

MariaAga commented:
@jeremylenz The alert is great there, while choosing to upgrade all and having no packages the review step is blocked with no explanation, can you please add an alert for that as well?

MariaAga commented on webpack/components/extensions/Hosts/BulkActions/BulkPackagesWizard/02_BulkPackagesTable.js:
```diff
@@ -0,0 +1,155 @@
+  };
+
+  // to pass in so that TableIndexPage doesn't alter browser URL
+  const restrictedSearchQuery = query => query;
```
Is this something other components will need to do?

jeremylenz replied:
Only components with "embedded" tables (in other words tables that you need to not affect the browser URL params).

MariaAga replied:
Should it be added as code/comment in the foreman?

jeremylenz replied:
Realized that it makes more sense to just use the `updateParamsByUrl` param which was already there. Updated here and in Foreman.

jeremylenz commented:
Also added the other alert in the case of no installed packages in the system.

jeremylenz commented:
Rebased to fix merge conflict. 
also removed the commit that alters tests since the Foreman PR is now merged.

parthaa reviewed (approved):
Works for me.

nadjaheitmann commented on webpack/scenes/Tasks/helpers.js:
```diff
@@ -2,7 +2,7 @@ import React from 'react';
 import store from 'foremanReact/redux';
 import { urlBuilder } from 'foremanReact/common/urlHelpers';
-import { addToast } from 'foremanReact/components/ToastsList/slice.js';
+import { addToast } from 'foremanReact/components/ToastsList/slice';
```
This (not fixed) line causes a bug in Katello 4.13, toasts fail for several REX tasks. Can we backport this one line to Katello 4.13? :) I can open a PR against the stable branch if you'd like :)

MariaAga replied:
Which error does it cause? if this causes an error wouldnt the other imports be a problem as well?

nadjaheitmann replied:
The error is caused if the line stays as is: 
`import { addToast } from 'foremanReact/components/ToastsList/slice.js';`
Not sure what exactly is causing the problem but the `.js` file name seems to cause issues. So whenever a REX job is run on 4.13, an empty error toast is dispatched.

![image](https://github.com/user-attachments/assets/134c4555-3a86-41dc-b310-84886eba5f17)

MariaAga replied:
Good to know! Probably will be good to open a pr to backport it, but thats for the Katello people to approve :)

jeremylenz replied:
Yes, please feel free to open a PR in KATELLO-4.13 :+1:

nadjaheitmann commented:
Hey @jeremylenz , we are trying to integrate Debian packages here as well and I am kind of confused by the functionality of the wizard. What happens if you select hosts with different OSes? Which packages is it supposed to show for "Install packages" ? I can see that it is possible to install packages with similar names on different OSes but otherwise the task will simply fail. 

Moreover, it gets complicated if you have debs and rpms because they have different links.

So I am not sure what the best way to go is:
We can either 1) look at the operatingsystem_family of the hosts and then remove install/upgrade single packages or 2) try fixing the links in a correct manner. 

It seems to me that the minimum that should be done is not showing rpms/deb when only hosts with the other OS family are selected.

What are your thoughts?

jeremylenz commented:
@nadjaheitmann good questions! The `thindex` endpoint I added only considers `Rpm` type when returning the list of possible packages to install. For update and remove, it uses installed packages, which I believe can include deb. 

I'm thinking the simplest way forward may be a separate wizard for Debs. It could use different (and new if needed) endpoints to display the package lists, but could probably use the same REX templates. An added benefit is that we could hide the deb (or rpm) wizard downstream if needed.

jeremylenz commented:
Or maybe add a different wizard step - so you could have "Install rpms" and "Install debs" as separate wizard paths? just brainstorming

sbernhard commented:
@jeremylenz This is not a debian-only issue - it gets even worse with debian. 
But the same issue happens if you, for example select hosts with (RHEL 7 and RHEL8) or (SLES12 and RHEL8) or other weird combinations.

jeremylenz commented:
I think that installing packages from multiple OSes should not be a design requirement of this wizard. If it needs more in-depth discussion, perhaps we can move to a post on community.theforeman.org

nadjaheitmann commented:
So what is the design requirement for the package installation/update/removal? Is the wizard still WIP like the new all hosts page? If you have multiple RHEL hosts, they can also have different usages and, hence, different packages/content views, etc. I am just trying to understand the big picture. I can open a discourse thread for more detailed discussion.

nadjaheitmann commented:
Here is the discourse post: https://community.theforeman.org/t/rfc-new-all-hosts-page-wizard-for-package-management/41996 

I would also like to repeat here: Thanks for the great effort and looking forward to contribute :)
